crates/adapters/src/transport/url.rs

```diff
@@ -0,0 +1,248 @@
+use super::{InputConsumer, InputEndpoint, InputTransport};
+use crate::PipelineState;
+use actix::System;
+use anyhow::{anyhow, Result as AnyResult};
+use awc::{Client, Connector};
+use futures::StreamExt;
+use lazy_static::lazy_static;
+use rustls::{ClientConfig, OwnedTrustAnchor, RootCertStore};
+use serde::Deserialize;
+use serde_yaml::Value as YamlValue;
+use std::{borrow::Cow, sync::Arc, thread::spawn};
+use tokio::{
+    select,
+    sync::watch::{channel, Receiver, Sender},
+};
+use utoipa::ToSchema;
+use webpki_roots::TLS_SERVER_ROOTS;
+
+/// [`InputTransport`] implementation that reads data from an HTTP or HTTPS URL.
+///
+/// The input transport factory gives this transport the name `url`.
+pub struct UrlInputTransport;
+
+impl InputTransport for UrlInputTransport {
+    fn name(&self) -> Cow<'static, str> {
+        Cow::Borrowed("url")
+    }
+
+    /// Creates a new [`InputEndpoint`] for reading from an HTTP or HTTPS URL,
+    /// interpreting `config` as a [`UrlInputConfig`].
+    ///
+    /// See [`InputTransport::new_endpoint()`] for more information.
+    fn new_endpoint(&self, _name: &str, config: &YamlValue) -> AnyResult<Box<dyn InputEndpoint>> {
+        let config = UrlInputConfig::deserialize(config)?;
+        let ep = UrlInputEndpoint::new(config);
+        Ok(Box::new(ep))
+    }
+}
+
+/// Configuration for reading data from an HTTP or HTTPS URL with
+/// [`UrlInputTransport`].
+#[derive(Clone, Deserialize, ToSchema)]
+pub struct UrlInputConfig {
+    /// URL.
+    pub path: String,
+}
+
+struct UrlInputEndpoint {
+    config: UrlInputConfig,
+    sender: Sender<PipelineState>,
+    receiver: Receiver<PipelineState>,
+}
+
+impl UrlInputEndpoint {
+    fn new(config: UrlInputConfig) -> Self {
+        let (sender, receiver) = channel(PipelineState::Paused);
+        Self {
+            config,
+            sender,
+            receiver,
+        }
+    }
+
+    async fn worker_thread(
+        config: UrlInputConfig,
+        consumer: &mut Box<dyn InputConsumer>,
+        mut receiver: Receiver<PipelineState>,
+    ) -> AnyResult<()> {
+        let client = Client::builder()
+            .connector(Connector::new().rustls(rustls_config()))
+            .finish();
+        let request = client.get(config.path);
+        let mut response = request.send().await.map_err(
+            // `awc` intentionally uses errors that aren't `Sync`, but
+            // `anyhow::Error` requires `Sync`.  Transform the error so we can
+            // return it.
+            |error| anyhow!("{error}"),
+        )?;
+
+        while *receiver.borrow() != PipelineState::Terminated {
+            select! {
+                _ = receiver.changed() => (),
+                result = response.next(), if *receiver.borrow() == PipelineState::Running => {
```

Comment:
I guess if we stay in the paused state long enough, the server will eventually drop the connection and we won't be able to continue. I guess there's not a lot we can do about it without adding too much complexity.

Reply:
I pushed a change to connect on un-pause, disconnect on pause, and reconnect but skip past data already read if un-paused again.